Stratis.Bitcoin/Base/ChainHeadersBehavior.cs

```diff
@@ -7,326 +7,390 @@
+            // This updates the pending tip parameter which is 
+            // the peers current best validated height.
+            // If the peer's height is higher Chain.Tip is updated to have 
+            // the most PoW header.
+            // It is sent in response to GetHeadersPayload or is solicited by the 
+            // peer when a new block is validated (and not in IBD).
+
+            var newHeaders = message.Message.Payload as HeadersPayload;
+            if ((newHeaders != null) && this.CanSync)
+            {
+                ChainedBlock pendingTipBefore = this.GetPendingTipOrChainTip();
+                this.logger.LogTrace("Pending tip is '{0}', received {1} new headers.", pendingTipBefore, newHeaders.Headers.Count);
+
+                // TODO: implement MAX_HEADERS_RESULTS in NBitcoin.HeadersPayload
+
+                ChainedBlock tip = pendingTipBefore;
+                foreach (BlockHeader header in newHeaders.Headers)
+                {
+                    ChainedBlock prev = tip.FindAncestorOrSelf(header.HashPrevBlock);
+                    if (prev == null)
+                        break;
+
+                    tip = new ChainedBlock(header, header.GetHash(), prev);
+                    bool validated = this.Chain.GetBlock(tip.HashBlock) != null || tip.Validate(this.AttachedNode.Network);
+                    validated &= !this.chainState.IsMarkedInvalid(tip.HashBlock);
+                    if (!validated)
+                    {
+                        this.logger.LogTrace("Validation of new header '{0}' failed.", tip);
+                        this.invalidHeaderReceived = true;
+                        break;
+                    }
+
+                    this.pendingTip = tip;
+                }
+
+                if (pendingTipBefore != this.pendingTip)
+                    this.logger.LogTrace("Pending tip changed to '{0}'.", this.pendingTip);
+
+                // Long reorganization protection on POS networks.
+                bool reorgPrevented = false;
+                uint maxReorgLength = this.chainState.MaxReorgLength;
+                Network network = this.AttachedNode?.Network;
+                ChainedBlock consensusTip = this.chainState.ConsensusTip;
+                if ((maxReorgLength != 0) && (network != null) && (consensusTip != null))
+                {
+                    ChainedBlock fork = this.pendingTip.FindFork(consensusTip);
+                    if ((fork != null) && (fork != consensusTip))
+                    {
+                        int reorgLength = consensusTip.Height - fork.Height;
+                        if (reorgLength > maxReorgLength)
+                        {
+                            this.logger.LogTrace("Reorganization of length {0} prevented, maximal reorganization length is {1}, consensus tip is '{2}'.", reorgLength, maxReorgLength, consensusTip);
+                            this.invalidHeaderReceived = true;
```

Comment:
I think we should trigger a disconnect from this node.
But perhaps we will do that when working on node banning (as we will also want to remember a banned node).

Reply:
yes this is for node banning